cellmlmanip/units.py

```diff
@@ -197,6 +197,13 @@ def add_base_unit(self, name):
         # Return new unit
         return getattr(self._registry, qname)
 
+    def add_conversion_rule(self, from_unit, to_unit, rule):
+        """Adds complex conversion rule for converting between incompatible units."""
+        context = pint.Context(str(from_unit) + str(to_unit) + str(rule))
+        context.add_transformation(from_unit, to_unit, rule)
+        self._registry.add_context(context)
+        self._registry.enable_contexts(str(from_unit) + str(to_unit) + str(rule))
```

Comment:
@jonc125  hhm yes it works but it you're still creating a new context every time and moving rules, which feels like it shouldn't be necessary.

Also disable_contexts() disables ALL contexts. Can we guarantee going forward that we won't want to use contexts elsewhere?

Reply:
Yeah, it's not ideal. I tried to figure out where/how pint was caching the context rules, but couldn't manage it. Even setting `context.checked=False` explicitly didn't help.

We could do `disable_contexts(1)` to disable just the last context added. But given the registry is private no-one should be adding contexts without using our API really...

Reply:
Or we could make `add_conversion_rule` just add rules to the context, and only enable the context the first time `get_conversion_factor` is called?

Reply:
Hmm, or maybe it's better to have a context per rule rather than relying on magic that might break, even if it is ugly? I'm conflicted!

Reply:
@jonc125 lol yes so was I when I wrote the initial version. It's a shame pints doesn't have a convenient way to do this.
In principle you're right the registry is private so they shouldn't be using it, but if in future we ourselves want to use context for something else (especially if it's been a while since working on this) we might inadvertently break it.

Enabling it upon first use would be fine, but we would still need a way to add new rules after. The reason I need this is exactly the scenario in the last test I've added:
The model has the stimulus_amplitude in the wrong units. the rule to convert uses the capacitance, but the capacitance is also in the wrong units, so needs to be converted before I can add the rule for the stimulus_emplitude.

So overall I still think a context rule per rule will be safest. At least for me I won't be adding more than a hand full of rules so performance wise I don't think it'll matter much.

Reply:
Yeah, let's go with that. Turns out that pint 0.9 that we have installed (as newer versions drop support for python 3.5) has `_checked` rather than `checked` as the attribute name, which might have been why that hack didn't work!

Reply:
makse sense and also probably indicate they don't want you touching it